### PR TITLE
use sccache on Windows

### PR DESCRIPTION
**Pull request by Be-ing**

### CODE REVIEW

**Comment by Be-ing** [comment]
Still not working after #3589...

**Comment by Be-ing** [comment]
Well this is strange. Commenting out the `/fp:fast` compiler flag gets rid of the `parse_arguments: CannotCache(multiple input files)` error from the sccache server, but now it fails with
```
..\lib\googletest-1.10.x\googletest\src\gtest-all.cc: fatal error C1041: cannot open program database 'D:\a\mixxx\mixxx\build\bin\gmock_main.pdb'; if multiple CL.EXE write to the same .PDB file, please use /FS
cl : Command line warning D9025 : overriding '/Z7' with '/Zi'
```
But there is no `/Zi` used?

**Comment by Be-ing** [comment]
Google Test sets the `Zi` compiler option which is incompatible with `Z7` and breaks the build. I submitted a [pull request upstream](https://github.com/google/googletest/pull/3291) to fix that.

**Comment by Holzhaus** [comment]
Maybe it's time to switch to doctest.

**Comment by Be-ing** [comment]
Requiring a tiny change to Google Test is IMO not a good reason to spend the effort switching away from it. But I am [pretty wary of relying on Goolge-maintained software](https://killedbygoogle.com/)...

**Comment by Be-ing** [comment]
Build step is down from [25 minutes](https://github.com/mixxxdj/mixxx/runs/1984559170) to fill the cache to [7 minutes](https://github.com/mixxxdj/mixxx/runs/1984717667) with cache hits. That brings the total Windows job time from 35 minutes to 19 minutes. That's a big improvement although still quite a bit slower than the ~10 minute jobs for Ubuntu and macOS.

**Comment by Be-ing** [comment]
I confirmed that the mixxx.pdb is still generated and installed with `/Z7`.

**Comment by Be-ing** [comment]
> Well this is strange. Commenting out the /fp:fast compiler flag gets rid of the parse_arguments: CannotCache(multiple input files) error from the sccache server

This was because of a bug in sccache which I have [fixed](https://github.com/mozilla/sccache/issues/950). GH Actions now builds my fixed sccache branch and caches the build.

**Comment by Be-ing** [comment]
> That brings the total Windows job time from 35 minutes to 19 minutes.

Using Ninja from the GH Actions VM image instead of downloading it from Chocolatey reduced the job time another 1.5-2 minutes, so now Windows jobs take 16-17 minutes. :rocket:

**Comment by daschuer** [comment]
What was the reason for switching from clcache to cscache? Maybe clcache did not work for another reason that is fixed now as well.
I am asking because of relying on a patched version of sccache is not optimal.

Is cmake able to verify if the sccache is patched?

Is the PDB debugging still possible? It was originsl broken with /Zi, but I have read somwher that Microsoft has fixed it. 

Please verify if this fix already works for us.

**Comment by Be-ing** [comment]
clcache is abandoned. Maybe it would work with the changes in this branch, but relying on unmaintained software is a bad idea. As far as I can tell, sccache is the only currently maintained compiler cache for MSVC.

I double checked and the `mixxx.pdb` file is still generated and installed with `/Z7`.

**Comment by Be-ing** [comment]
> Is cmake able to verify if the sccache is patched?

No, but the build still works with the unpatched sccache. It just doesn't cache.

**Comment by daschuer** [comment]
If clcache just works for us, there is no need for active development. I would prefer to use this, because it will introduce less hassle to a new contributor, compared to use a patched version of cscache. It is even more bad if other believe to use it, but waiting ages anyway because of the bug.

**Comment by Be-ing** [comment]
Installing the patched sccache is no harder than clcache. Once the user has Rust installed, it is just:
```
cargo install --git https://github.com/Be-ing/sccache.git --branch fix_msvc_fp
```
and cargo does the rest.

> If clcache just works for us

But it doesn't, we don't know why, and I'm not wasting my time figuring out why unmaintained software doesn't work.

**Comment by Be-ing** [comment]
All that will change when my PR for sccache is merged is the cargo command:
```
cargo install sccache
```

**Comment by Be-ing** [comment]
I am opposed to adding dependencies on unmaintained tools when there is a currently maintained alternative which is used by large projects (Firefox and Rustc).

**Comment by Be-ing** [comment]
Merge?

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -86,8 +86,8 @@ jobs:
             # TODO: Fix these broken tests on Windows
             ctest_args: --exclude-regex '^AutoDJProcessorTest.*$'
             cpack_generator: WIX
-            compiler_cache: clcache
-            compiler_cache_path: ${{ github.workspace }}\clcache
+            compiler_cache: sccache
+            compiler_cache_path: C:\Users\runneradmin\AppData\Local\Mozilla\sccache\cache
```
Comment (uklotzde) [inline .github/workflows/build.yml:90]: Did you try the environment variable LOCALAPPDATA instead of the hard-coded absolute path?

Comment (Be-ing) [reply .github/workflows/build.yml:90]: Yes. It did not work.

Comment (Be-ing) [reply .github/workflows/build.yml:90]: When I tried that, the cache restoration hit the key but the restored cache was empty.

File: .github/workflows/build.yml [hunk 2]
```diff
@@ -103,6 +103,25 @@ jobs:
...
     steps:
+
+    # sccache's handling of the /fp:fast MSVC compiler option is broken, so use our fork with the fix.
+    # https://github.com/mozilla/sccache/issues/950
+    - name: "[Windows] Set up cargo cache"
+      if: runner.os == 'Windows'
+      uses: actions/cache@v2
+      id: sccache-build-cache
+      with:
+        path: C:\Users\runneradmin\.cargo
```
Comment (uklotzde) [inline .github/workflows/build.yml:114]: %USERPROFILE% = C:\Users\runneradmin?

Not much information only found this source: https://www.theserverside.com/blog/Coffee-Talk-Java-News-Stories-and-Opinions/environment-variables-full-list-github-actions

File: .github/workflows/build.yml [hunk 3]
```diff
@@ -103,6 +103,25 @@ jobs:
...
+        # hash of commit to build
+        key: e12434e4db24ccbe52b83fb6e37afcb94fc8d37b
+
+    # This needs to be done first because something later messes with $PATH in a way that breaks cargo
+    # by causing it to find link.exe from Git for Windows instead of MSVC.
+    - name: "[Windows] Build fixed sccache"
+      shell: bash
+      if: runner.os == 'Windows' && steps.sccache-build-cache.outputs.cache-hit != 'true'
+      run: cargo install --git https://github.com/Be-ing/sccache.git --branch fix_msvc_fp
```
Comment (uklotzde) [inline .github/workflows/build.yml:123]: Please add a comment why this patched version is needed and a TODO that explains how to replace it later (including a link to the PR).

Comment (Be-ing) [reply .github/workflows/build.yml:123]: The comment is on the above step. I can move it here.

File: .github/workflows/build.yml (hunk 2 above)
Comment (Be-ing) [reply .github/workflows/build.yml:114]: As noted above, GH Actions caching doesn't seem work with Windows environment variables in the path.

Comment (Be-ing) [reply .github/workflows/build.yml:114]: I will not bother experimenting with that further. It works how it is and the absolute path is unlikely to ever change.

File: .github/workflows/build.yml (hunk 3 above)
Comment (Be-ing) [reply .github/workflows/build.yml:123]: I am torn whether to move the comment down here or leave it where it is. What do you think?

Comment (Be-ing) [reply .github/workflows/build.yml:124]: > But including the final version of the cargo install command as a comment would still be helpful.

This is IMO obvious from normal usage of cargo and the sccache documentation, but I added a comment anyway.

**Review by uklotzde — APPROVED** [review]
Thanks for pushing this even further! LGTM

**Comment by daschuer** [comment]
We have now a 9m 29s penalty for build sccache from source. 
I think we should fetch a pre build binary once the patch is merged.

**Comment by Holzhaus** [comment]
> We have now a 9m 29s penalty for build sccache from source. 
> I think we should fetch a pre build binary once the patch is merged.

https://github.com/actions/cache/blob/master/examples.md#rust---cargo

**Comment by Be-ing** [comment]
The sccache build is already cached. GH Actions separates caches by branch so it may need to be rebuilt for the initial merge. But now it should take < 1 minute to restore from the cache.

Using the hash of the Cargo.lock file would be a good idea. Currently the key is just the commit hash of sccache Git repository. Of course that won't scale when we use cargo for more binaries such as Aoide.

**Pull request merged**